pkg/envoy/registry/services.go

```diff
@@ -0,0 +1,98 @@
+package registry
+
+import (
+	"fmt"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/openservicemesh/osm/pkg/envoy"
+	k8s "github.com/openservicemesh/osm/pkg/kubernetes"
+	"github.com/openservicemesh/osm/pkg/service"
+)
+
+// ProxyServiceMapper knows how to map Envoy instances to services.
+type ProxyServiceMapper interface {
+	ListProxyServices(*envoy.Proxy) ([]service.MeshService, error)
+}
+
+// ExplicitProxyServiceMapper is a custom ProxyServiceMapper implementation.
+type ExplicitProxyServiceMapper func(*envoy.Proxy) ([]service.MeshService, error)
+
+// ListProxyServices executes the given mapping.
+func (e ExplicitProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service.MeshService, error) {
+	return e(p)
+}
+
+// KubeProxyServiceMapper maps an Envoy instance to services in a Kubernetes cluster.
+type KubeProxyServiceMapper struct {
+	KubeController k8s.Controller
+}
+
```

Comment:
I am not in favor of exposing k8s constructs under `pkg/envoy`. We have tried to get away from this as much as possible (ex. service account -> service identity changes).

Is this avoidable?

Reply:
That makes sense. I can't say for certain it's unavoidable, I'll see if I can refactor this.

Reply:
@shashankram What would you recommend? I am curious what you have in mind?

I agree on keeping k8s out of `pkg/envoy`, but also it'd be tough to determine services for envoy (pod) without leveraging k8s.  So this boils down to perhaps a different package (location) where the `ProxyServiceMapper` implementation lives?

I would not be opposed to @nojnhuh doing this in another PR if he prefers to do that.

Reply:
@nojnhuh and I chatted a bit about this. None of my comments need to be resolved in this specific PR, though @nojnhuh wanted to play around with refactoring the code a bit more to explore the feasibility of my comments.

Reply:
> I am not in favor of exposing k8s constructs under pkg/envoy.

Having played with this a bit, it seems like if this is moved into pkg/kubernetes, then we're exposing Envoy details there which doesn't seem any better. I suppose we could create another package for this, but that doesn't really solve the problem of mixing kuberentes and envoy details.

To really fix this it seems like we'd need a set of implementation-agnostic proxy abstractions so this could then map kubernetes services to abstract proxies instead of Envoy proxies.